### PR TITLE
feat: add support for hostname-based DRS URIs

### PR DESCRIPTION
**Pull request by sarthakgupta072**
* [Hostname-based DRS URIs](https://ga4gh.github.io/data-repository-service-schemas/preview/develop/docs/#_hostname_based_drs_uris) can be passed to both constructor and endpoint access methods

### CODE REVIEW

File: drs_cli/client.py [hunk 1]
```diff
@@ -30,18 +33,27 @@ class DRSClient():
     def __init__(
         self,
         host: str = 'http://0.0.0.0',
-        port: int = '8080',
+        port: int = 80,
         base_path: str = 'ga4gh/drs/v1',
+        drs_uri: Optional[str] = None,
```
Comment (uniqueg) [inline drs_cli/client.py]: From the user's perspective, it might be easier to have the first parameter (you could call it 
`uri`) take either the API host _or_ a DRS URI. As the port is usually 80 or 443 (you can set that based on whether the user supplies a host with `http` or `https`, respectively), and the base path is usually the same (as it is defined in the API specs), the user could call the constructor in most cases just with a single positional parameter, passing either host or DRS URI, e.g., `DRSClient("drs://my-drs.org/A3D4SR")` or `DRSClient("https://my-drs.org")`. You can then call your `_check_drs_uri_regex()` and if that fails, call a `_check_http_uri_regex()` and if both fail, you give an error message that says that the passed connection cannot be interpreted and that either a HTTP(S) or DRS URI needs to be passed.

File: drs_cli/client.py [hunk 2]
```diff
@@ -30,18 +33,27 @@ class DRSClient():
...
         token: Optional[str] = None,
-    ) -> None:
+    ) -> Union[None, Error]:
+        
         self.url = f"{host}:{port}/{base_path}"
+        if drs_uri:
+            if _check_drs_regex(drs_uri):
+                drs_host = _get_host_and_id(drs_uri)[0]
+                self.url = f"http://{drs_host}/{base_path}"
```
Comment (uniqueg) [inline drs_cli/client.py]: I think you can refactor the code to assign to `self.uri` only once, at the end. I would also cover for the port problem when passing a DRS URI: I would set the default port to 443 instead of 80. Then, when the user passes a a HTTP (and not HTTPS) URL but not a port, then change it to 80. But if the user supplies a port, then use the provided port, even for DRS URIs. This is fine as we are not breaking the specs, the DRS URI does not contain the port, we provide it manually. In this way we can support the case where we don't have HTTPS DRS instances during development and/or are deploying to different ports.

File: drs_cli/client.py [hunk 3]
```diff
@@ -30,18 +33,27 @@ class DRSClient():
...
+                return Error(**DRS_URI_FORMAT_ERROR)
         self.token = token
         self._get_headers()
         logger.info(f"API URL: {self.url}")
 
     def get_object(
         self,
-        object_id: str,
+        object_id: Optional[str] = None,
+        drs_uri: Optional[str] = None,
```
Comment (uniqueg) [inline drs_cli/client.py]: Same thing as before: just use `object` which can be either an object ID or a DRS URI, for the user's convenience. Then do the regex checks (one for DRS URI and one for allowed object IDs, as defined in the DRS specs) to determine which was passed

File: drs_cli/client.py [hunk 4]
```diff
@@ -193,3 +213,15 @@ def _get_headers(self):
         }
         if self.token:
             self.headers['Authorization'] = 'Bearer ' + self.token
+
+
+def _check_drs_regex(uri) -> bool:
```
Comment (uniqueg) [inline drs_cli/client.py]: I would probably merge the two methods to check and parse into a single method that parses and returns host and object if it can and raises an error otherwise. Then in the main code call the function and assign to a tuple as you do for the `_get_host_and_id()` function, but wrap it in a `try`/`except` block

**Review by uniqueg — CHANGES_REQUESTED** [review]
Definitely on the right path, but couple of things to make it easier for the user and make the code a bit easier

File: drs_cli/client.py (hunk 3 above)
Comment (sarthakgupta072) [reply drs_cli/client.py]: Should I leave this as it was before, coz I have added regex for HTTP, HTTPS URIs, and also regex for DRS IDs. It improves the readability of the code.

File: drs_cli/client.py (hunk 4 above)
Comment (sarthakgupta072) [reply drs_cli/client.py]: Should I leave this as it was before (two separate methods), coz I have added regex for HTTP, HTTPS URIs, and also regex for DRS IDs? It improves the readability of the code.

Comment (sarthakgupta072) [reply drs_cli/client.py]: And the `if` conditions are easier to write when the methods are separate. 😅

File: drs_cli/client.py [hunk 5]
```diff
@@ -1,229 +1,424 @@
...
-        token: Bearer token for authentication.
+        uri: URI to DRS endpoints, built from `uri`, `port` and `base_path`,
+            e.g.,"https://my-drs.app:443/ga4gh/drs/v1".
+        token: Bearer token for gaining access to DRS endpoints.
         headers: Dictionary of request headers.
     """
+    # set regular expressions as private class variables
+    _RE_DOMAIN_PART = r'[a-z0-9]([a-z0-9-]{1,61}[a-z0-9])?'
+    _RE_DOMAIN = rf"({_RE_DOMAIN_PART}\.)+{_RE_DOMAIN_PART}\.?"
+    _RE_DRS_ID = r'.+'
```
Comment (sarthakgupta072) [inline drs_cli/client.py:61]: If you see this https://ga4gh.github.io/data-repository-service-schemas/preview/develop/docs/#_drs_ids

It says:
> DRS IDs are strings made up of uppercase and lowercase letters, decimal digits, hyphen, period, underscore and tilde [A-Za-z0-9.-_~]. See RFC 3986 § 2.3.

But I think this will contain any character for DRS_IDs right?

File: drs_cli/client.py [hunk 6]
```diff
@@ -1,229 +1,424 @@
...
 
     def get_object(
         self,
-        drs_object: str,
+        object_id: str,
         token: Optional[str] = None,
-    ) -> Union[Error, DrsObject]:
-        """Access DRS object.
+    ) -> Optional[Union[Error, DrsObject]]:
+        """Retrieve DRS object.
```
Comment (sarthakgupta072) [inline drs_cli/client.py:89]: Why is `Optional` added here?

File: drs_cli/client.py [hunk 7]
```diff
@@ -1,229 +1,424 @@
...
-            pydantic.ValidationError: The returned response does not conform
-                to the models defined in the API specification.
+            requests.exceptions.ConnectionError: A connection to the provided
+                DRS instance could not be established.
+            drs_cli.errors.InvalidResponseError: The response could not be
+                validated against the API schema.
         """
-        request_url = f"{self.url}/objects/{object_id}/access/{access_id}"
+        obj_id = self._get_object_id(object_id=object_id)
+        acc_id = quote(string=access_id, safe='')
```
Comment (sarthakgupta072) [inline drs_cli/client.py:179]: I think `quote` is not required here, if the IDs don't contain special characters as per https://ga4gh.github.io/data-repository-service-schemas/preview/develop/docs/#_drs_ids.

Also `quote` doc here https://docs.python.org/3/library/urllib.parse.html#urllib.parse.quote says:
> Replace special characters in a string using the %xx escape. Letters, digits, and the characters **'_.-~'** are never quoted.

which are all the characters that follow the DRS specs.

File: drs_cli/client.py [hunk 8]
```diff
@@ -1,229 +1,424 @@
...
-from drs_cli.errors import DRS_URI_FORMAT_ERROR
+from drs_cli.errors import (
+    InvalidObjectData, exception_handler,
+    InvalidObjectIdentifier,
+    InvalidResponseError,
+    InvalidURI,
+)
 
 logger = logging.getLogger(__name__)
+sys.excepthook = exception_handler
```
Comment (sarthakgupta072) [inline drs_cli/client.py:23]: Why is this required? 😅

File: drs_cli/client.py (hunk 5 above)
Comment (uniqueg) [reply drs_cli/client.py:61]: Yes, correct. But if you read on, it says:
> DRS IDs can contain other characters, but they MUST be encoded into valid DRS IDs whenever they are used in API calls. This is because non-encoded IDs may interfere with the interpretation of the objects/{id}/access endpoint. To overcome this limitation use percent-encoding of the ID, see RFC 3986 § 2.4

This is why I am not imposing any rules for characters after the slash following the hostname (`drs://host.name/<#!HERE_BE_DRAGONS#!>`). Instead, I have added some lines to percent-encode the ID as suggested (the `quote(string='xyz', safe='')` stuff).

File: drs_cli/client.py (hunk 4 above)
Comment (uniqueg) [reply drs_cli/client.py]: I implemented this now such that checking and parsing is done in one go, though separately for extracting the host and the object ID.

File: drs_cli/client.py (hunk 8 above)
Comment (uniqueg) [reply drs_cli/client.py:23]: I didn't like that the program prints a traceback whenever any exception occurs. Exceptions that are raised for clearly defined reasons (e.g., the user enters a wrong URL etc.) should not print a traceback, just some informative error message. To make sure we are in full control of how errors are handled I have defined our own error handler

File: drs_cli/client.py (hunk 6 above)
Comment (uniqueg) [reply drs_cli/client.py:89]: Good catch - it isn't! When I was working on the changes yesterday, I was switching between returning a Tuple of `(None, None)` and raising errors. Eventually I settled for raising errors, but I forgot to remove this, so please go ahead and remove it in your next commit.

File: drs_cli/client.py (hunk 7 above)
Comment (uniqueg) [reply drs_cli/client.py:179]: I know, but as I said before: DRS IDs _can_ contain other characters, in which case they should be percent-encoded when used in API calls.

File: drs_cli/client.py [hunk 9]
```diff
@@ -1,195 +1,424 @@
...
+           ('https', 'my-drs.app')
+           >>> DRSClient.get_host(uri="drs://my-drs.app/My0bj3ct")
+           ('drs', 'my-drs.app')
+        """
+        match = re.search(self._RE_HOST, uri, re.I)
+        if match:
+            schema = match.group('schema')
+            host = match.group('host').rstrip('\\')
+            if len(host) > 253:
+                raise InvalidURI
```
Comment (sarthakgupta072) [inline drs_cli/client.py:398]: This condition won't happen ever, right?

Because above the regex is `_RE_DOMAIN_PART = r'[a-z0-9]([a-z0-9-]{1,61}[a-z0-9])?'` which allows only 61 charater repetition?

Comment (uniqueg) [reply drs_cli/client.py:398]: Note that the capture group `host` is defined by `_RE_DOMAIN = rf"({_RE_DOMAIN_PART}\.)+{_RE_DOMAIN_PART}\.?"` which contains multiple repetitions of `_RE_DOMAIN_PART`. So yeah, each part (like `google` and `com` in `google.com`) can be at most 63 characters (first and last character cannot be hyphens), but all together cannot exceed 253 chars.

File: drs_cli/client.py [hunk 10]
```diff
@@ -1,195 +1,428 @@
...
+
+        Raises:
+            drs_cli.errors.InvalidObjectIdentifier: input object ID cannot be
+                parsed.
+        """
+        match = re.search(self._RE_OBJECT_ID, object_id, re.I)
+        try:
+            return quote(string=match.group('obj_id'), safe='')
+        except Exception:
+            raise InvalidObjectIdentifier
```
Comment (sarthakgupta072) [inline drs_cli/client.py]: This single line is only left from achieving `100%` coverage. I tried several test cases, but not able to make `quote` raise an error. :/

**Review by uniqueg — APPROVED** [review]
